google/datalab/contrib/pipeline/_pipeline.py

```diff
@@ -182,48 +163,123 @@ def _get_dependency_definition(task_id, dependencies):
-      type need to be made.
-    """
-    if (operator_type == 'bq'):
-      if (param_name == 'query'):
-        return 'bql'
-    return param_name
-
-  def _get_operator_param_value(self, param_name, operator_type, param_value):
-    """ Internal helper gets the python parameter value for the Airflow
-      operator class. It needs to make exceptions that are specific to the
-      operator-type, in some ways similar to _get_operator_param_name.
+  def _get_operator_param_name_and_values(operator_class_name, task_details):
+    """ Internal helper gets the name of the python parameter for the Airflow operator class. In
+      some cases, we do not expose the airflow parameter name in its native form, but choose to
+      expose a name that's more standard for Datalab, or one that's more friendly. For example,
+      Airflow's BigQueryOperator uses 'bql' for the query string, but we want %%bq users in Datalab
+      to use 'query'. Hence, a few substitutions that are specific to the Airflow operator need to
+      be made.
+
+      Similarly, we the parameter value could come from the notebook's context. All that happens
+      here.
     """
-    if (operator_type == 'bq') and (param_name in ['query', 'bql']):
-        return param_value.sql
-    return param_value
+
+    # We make a clone and then remove 'type' and 'up_stream' since these aren't needed for the
+    # the operator's parameters.
+    operator_task_details = task_details.copy()
+    if 'type' in operator_task_details.keys():
+      del operator_task_details['type']
+    if 'up_stream' in operator_task_details.keys():
+      del operator_task_details['up_stream']
+
+    if (operator_class_name == 'BigQueryOperator'):
+      return Pipeline._get_bq_execute_params(operator_task_details)
+    if (operator_class_name == 'BigQueryToCloudStorageOperator'):
+      return Pipeline._get_bq_extract_params(operator_task_details)
+    if (operator_class_name == 'GoogleCloudStorageToBigQueryOperator'):
+      return Pipeline._get_bq_load_params(operator_task_details)
+    return operator_task_details
 
   @staticmethod
-  def _add_default_override_params(task_details, operator_type):
-    """ Internal helper that overrides the defaults of an Airflow operator's
-      parameters, when necessary.
-    """
-    if operator_type == 'bq':
-      bq_defaults = {}
-      bq_defaults['use_legacy_sql'] = False
-      for param_name, param_value in bq_defaults.items():
-          if param_name not in task_details:
-            task_details[param_name] = param_value
+  def _get_bq_execute_params(operator_task_details):
+    if 'query' in operator_task_details:
+      operator_task_details['bql'] = operator_task_details['query'].sql
+      del operator_task_details['query']
+
+    # Add over-rides of Airflow defaults here.
+    if 'use_legacy_sql' not in operator_task_details:
+      operator_task_details['use_legacy_sql'] = False
```

Comment:
Datalab only supports GoogleSQL for BigQuery, so should force use_legacy_sql to always be false?

Reply:
Is that a question about why it's false here? Or is that a statement about it needing to be True?

Airflow makes this True by default:
https://github.com/apache/incubator-airflow/blob/master/airflow/contrib/operators/bigquery_operator.py#L68

And Datalab makes this False by default:
https://github.com/googledatalab/pydatalab/blob/b27dc582882fe5497af49d45fc28204e4c46db12/google/datalab/bigquery/_api.py#L170

We need to be consistent with the rest of Datalab here. Hence, we need to override the Airflow default (of True) and set it to False.